include/bluetooth/mesh/time_srv.h

```diff
@@ -0,0 +1,192 @@
+ */
+#define BT_MESH_TIME_SRV_INIT                                                  \
+	{                                                                      \
+		.pub = { .msg = NET_BUF_SIMPLE(BT_MESH_MODEL_BUF_LEN(          \
+				 BT_MESH_TIME_OP_TIME_SET,                     \
+				 BT_MESH_TIME_MSG_LEN_TIME_SET)) },            \
+		.setup_pub = {                                                 \
+			.msg = NET_BUF_SIMPLE(BT_MESH_MODEL_BUF_LEN(           \
+				BT_MESH_TIME_OP_TIME_ROLE_STATUS,              \
+				BT_MESH_TIME_MSG_LEN_TIME_ROLE_STATUS))        \
+		},                                                             \
+	}
+
+/** @def BT_MESH_MODEL_TIME_SRV
+ *
+ * @brief Time Server model composition data entry.
+ *
+ * @param[in] _srv Pointer to a @ref bt_mesh_time_srv instance.
+ */
+#define BT_MESH_MODEL_TIME_SRV(_srv)                                        \
+	BT_MESH_MODEL_CB(                                                    \
+		BT_MESH_MODEL_ID_TIME_SRV, _bt_mesh_time_srv_op, &(_srv)->pub, \
+		BT_MESH_MODEL_USER_DATA(struct bt_mesh_time_srv, _srv),        \
+		&_bt_mesh_time_srv_cb),                                        \
+		BT_MESH_MODEL_CB(BT_MESH_MODEL_ID_TIME_SETUP_SRV,              \
+				 _bt_mesh_time_setup_srv_op,                   \
+				 &(_srv)->setup_pub,                           \
+				 BT_MESH_MODEL_USER_DATA(                      \
+					 struct bt_mesh_time_srv, _srv),       \
+				 NULL)
+
+/** Internal data structure of the Time Server instance. */
+struct bt_mesh_time_srv_data {
+	/** Current TAI Seconds state. */
+	u64_t tai_sec;
+	/** TAI time at last recieved update. */
+	u64_t tai_sec_last_update;
+	/** Fractional part of TAI second in ms. */
+	u16_t subsec_ms;
+	/** Base part of uncertainty. */
+	u8_t uncertainty_base;
+	/** Local drift part of uncertainty. */
+	u8_t uncertainty_drift;
+	/** Reliable TAI source flag. */
+	bool is_authority;
+	/** Current Time Zone Offset. */
+	u8_t time_zone_offset_current;
+	/** Upcomming Time Zone Offset. */
+	u8_t time_zone_offset_new;
+	/** TAI update point for Time Zone Offset. */
+	u64_t tai_of_zone_change;
+	/** Current TAI-UTC Delta. */
+	u16_t tai_utc_delta_current;
+	/** Upcomming TAI-UTC Delta. */
+	u16_t tai_utc_delta_new;
+	/** TAI update point for TAI-UTC Delta. */
+	u64_t tai_of_delta_change;
+	/** The Time Role of the Server instance*/
+	enum time_role role;
+	/** Device uptime at last sync point */
+	s64_t sync_uptime;
+};
+
+/**
+ * Time Server instance. Should be initialized with
+ * @ref BT_MESH_TIME_SRV_INIT.
+ */
+struct bt_mesh_time_srv {
+	/** Model entry. */
+	struct bt_mesh_model *model;
+	/** Publish parameters. */
+	struct bt_mesh_model_pub pub;
+	/** Setup model publish parameters */
+	struct bt_mesh_model_pub setup_pub;
+	/** Acknowledged message tracking. */
+	struct bt_mesh_model_ack_ctx ack_ctx;
+	/** Model state structure */
+	struct bt_mesh_time_srv_data data;
+};
```

Comment:
I've added a optional CB that passes the entire server instance for now since I also are uncertain on this topic. We may discuss what is reasonable to pass to the user later, because i suspect that the current version might be a bit excessive.